cookbooks/tile/recipes/default.rb

```diff
@@ -522,77 +522,134 @@
-end
+  systemd_service "expire-tiles" do
+    description "Tile dirtying service"
+    type "simple"
+    user "_renderd"
+    exec_start "/usr/local/bin/expire-tiles"
+    nice 10
+    sandbox true
+    read_write_paths tile_directories + [
+                       "/store/database/nodes",
+                       "/var/lib/replicate/expire-queue",
+                       "/var/log/tile"
+                     ]
+  end
 
-service "expire-tiles.path" do
-  action [:enable, :start]
-  subscribes :restart, "systemd_path[expire-tiles]"
-end
+  systemd_path "expire-tiles" do
+    description "Tile dirtying trigger"
+    directory_not_empty "/var/lib/replicate/expire-queue"
+  end
 
-systemd_service "replicate" do
-  description "Rendering database replication service"
-  after "postgresql.service"
-  wants "postgresql.service"
-  user "tile"
-  exec_start "/usr/local/bin/replicate"
-  sandbox :enable_network => true
-  restrict_address_families "AF_UNIX"
-  read_write_paths [
-    "/store/database/nodes",
-    "/var/lib/replicate",
-    "/var/log/tile"
+  service "expire-tiles.path" do
+    action [:enable, :start]
+    subscribes :restart, "systemd_path[expire-tiles]"
+  end
+
+  systemd_service "replicate" do
+    description "Rendering database replication service"
+    after "postgresql.service"
+    wants "postgresql.service"
+    user "tile"
+    exec_start "/usr/local/bin/replicate"
+    sandbox :enable_network => true
+    restrict_address_families "AF_UNIX"
+    read_write_paths [
+      "/store/database/nodes",
+      "/var/lib/replicate",
+      "/var/log/tile"
+    ]
+    restart "on-failure"
+  end
+
+  service "replicate" do
+    action [:enable, :start]
+    subscribes :restart, "template[/usr/local/bin/replicate]"
+    subscribes :restart, "systemd_service[replicate]"
+  end
+elsif node[:tile][:replication][:engine] == "osm2pgsql"
+  service "expire-tiles.path" do
+    action [:disable, :stop]
+  end
+
+  systemd_path "expire-tiles" do
+    action :delete
+  end
+
+  service "expire-tiles" do
+    action :stop
+  end
+
+  systemd_service "expire-tiles" do
+    action :delete
+  end
+
+  template "/usr/local/bin/expire-tiles" do
+    source "expire-tiles-osm2pgsql.erb"
+    owner "root"
+    group "root"
+    mode "755"
+  end
+
+  osm2pgsql_arguments = %w[
+    --number-processes=1
+    --log-progress=false
+    --expire-tiles=13-19
+    --expire-output=/var/lib/replicate/dirty-tiles.txt
   ]
-  restart "on-failure"
-end
 
-service "replicate" do
-  action [:enable, :start]
-  subscribes :restart, "template[/usr/local/bin/replicate]"
-  subscribes :restart, "systemd_service[replicate]"
+  osm2pgsql_arguments.append("--multi-geometry") if node[:tile][:database][:multi_geometry]
+  osm2pgsql_arguments.append("--hstore") if node[:tile][:database][:hstore]
+  osm2pgsql_arguments.append("--tag-transform-script=#{node[:tile][:database][:tag_transform_script]}") if node[:tile][:database][:tag_transform_script]
+
```

Comment:
Looking at the docs and https://github.com/openstreetmap/operations/issues/987#issuecomment-1816047307 we'll need to specify the --style file.

Reply:
The `--style` does seem to be in the properties table?

```
gis=# select * from osm2pgsql_properties;
       property        |                                value                                 
-----------------------+----------------------------------------------------------------------
 attributes            | false
 db_format             | 2
 flat_node_file        | /store/database/nodes
 import_timestamp      | 2023-10-02T00:00:02Z
 output                | pgsql
 prefix                | planet_osm
 replication_timestamp | 2023-10-02T00:00:02Z
 style                 | /srv/tile.openstreetmap.org/styles/default/openstreetmap-carto.style
 updatable             | true
 version               | 1.9.2
 current_timestamp     | 2023-11-22T14:51:59Z
(11 rows)
```

Reply:
Oh, I guess it's fine then. I guess technically --style is present on the flex backend, although it means something different there.